### ISSUE TITLE
sql: cockroach CTE query result differs from postgres

### ISSUE THREAD

**Issue by mattcrdb** [cockroachdb/cockroach#42621]
**Describe the problem**

A customer observed the following behavior in a CTE query.

**To Reproduce**

CockroachDB:

```
CREATE TABLE pathdata (
bucket BYTES NOT NULL,
fullpath BYTES NOT NULL,
metadata BYTES NOT NULL,
CONSTRAINT "primary" PRIMARY KEY (bucket ASC, fullpath ASC),
FAMILY "primary" (bucket, fullpath, metadata),
CONSTRAINT check_fullpath CHECK (fullpath != '')
)

INSERT INTO pathdata (bucket, fullpath, metadata) VALUES (b'1'::BYTEA, b'2'::BYTEA, b'3'::BYTEA);

WITH updated AS (
DELETE FROM pathdata
WHERE pathdata.metadata = b'3':::BYTEA
AND pathdata.bucket = b'1':::BYTEA
AND pathdata.fullpath = b'2':::BYTEA
RETURNING 1
)
SELECT EXISTS(SELECT 1 FROM pathdata WHERE pathdata.bucket = b'1':::BYTEA AND pathdata.fullpath = b'2':::BYTEA) AS key_present,
EXISTS(SELECT 1) AS value_updated;


  key_present | value_updated
+-------------+---------------+
     false    |     true
(1 row)

EXPLAIN WITH updated AS (
DELETE FROM pathdata
WHERE pathdata.metadata = b'3':::BYTEA
AND pathdata.bucket = b'1':::BYTEA
AND pathdata.fullpath = b'2':::BYTEA
RETURNING 1
)
SELECT EXISTS(SELECT 1 FROM pathdata WHERE pathdata.bucket = b'1':::BYTEA AND pathdata.fullpath = b'2':::BYTEA) AS key_present,
EXISTS(SELECT 1 FROM updated) AS value_updated
;
                      tree                     |    field     |                                                                       description
+----------------------------------------------+--------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
                                               | distributed  | false
                                               | vectorized   | false
  root                                         |              |
   ├── values                                  |              |
   │                                           | size         | 2 columns, 1 row
   ├── subquery                                |              |
   │    │                                      | id           | @S1
   │    │                                      | original sql | DELETE FROM pathdata WHERE ((pathdata.metadata = b'3':::BYTES) AND (pathdata.bucket = b'1':::BYTES)) AND (pathdata.fullpath = b'2':::BYTES) RETURNING 1
   │    │                                      | exec mode    | all rows
   │    └── buffer node                        |              |
   │         │                                 | label        | buffer 1 (updated)
   │         └── spool                         |              |
   │              └── render                   |              |
   │                   └── run                 |              |
   │                        └── delete         |              |
   │                             │             | from         | pathdata
   │                             │             | strategy     | deleter
   │                             └── render    |              |
   │                                  └── scan |              |
   │                                           | table        | pathdata@primary
   │                                           | spans        | /"1"/"2"-/"1"/"2"/#
   │                                           | filter       | metadata = '\x33'
   ├── subquery                                |              |
   │    │                                      | id           | @S2
   │    │                                      | original sql | EXISTS (SELECT 1 FROM pathdata WHERE (pathdata.bucket = b'1':::BYTES) AND (pathdata.fullpath = b'2':::BYTES))
   │    │                                      | exec mode    | exists
   │    └── scan                               |              |
   │                                           | table        | pathdata@primary
   │                                           | spans        | /"1"/"2"-/"1"/"2"/#
   └── subquery                                |              |
        │                                      | id           | @S3
        │                                      | original sql | EXISTS (SELECT 1 FROM updated)
        │                                      | exec mode    | exists
        └── scan buffer node                   |              |
                                               | label        | buffer 1 (updated)
```

postgres behavior:

```
CREATE TABLE pathdata (
bucket BYTEA NOT NULL,
fullpath BYTEA NOT NULL,
metadata BYTEA NOT NULL
)

INSERT INTO pathdata (bucket, fullpath, metadata) VALUES ('1'::BYTEA, '2'::BYTEA, '3'::BYTEA);

matthewvardi=# WITH
	matching_key AS (SELECT * FROM pathdata WHERE bucket = '1'::BYTES AND fullpath = '2'::BYTES),
	updated AS (DELETE FROM pathdata WHERE pathdata.metadata = '3'::BYTES AND pathdata.bucket = '1'::BYTES AND pathdata.fullpath = '2'::BYTES RETURNING 1)
SELECT
	EXISTS(SELECT 1 FROM matching_key) AS key_present, EXISTS(SELECT 1 FROM updated) AS value_updated;
 key_present | value_updated 
-------------+---------------
 t           | t


                                                    QUERY PLAN                                                    
------------------------------------------------------------------------------------------------------------------
 Result  (cost=41.15..41.16 rows=1 width=2)
   CTE updated
     ->  Delete on pathdata  (cost=0.00..21.38 rows=1 width=6)
           ->  Seq Scan on pathdata  (cost=0.00..21.38 rows=1 width=6)
                 Filter: ((metadata = '\x33'::bytea) AND (bucket = '\x31'::bytea) AND (fullpath = '\x32'::bytea))
   InitPlan 2 (returns $2)
     ->  Seq Scan on pathdata pathdata_1  (cost=0.00..19.75 rows=1 width=0)
           Filter: ((bucket = '\x31'::bytea) AND (fullpath = '\x32'::bytea))
   InitPlan 3 (returns $3)
     ->  CTE Scan on updated  (cost=0.00..0.02 rows=1 width=0)
(10 rows)



```
**Expected behavior**
Query behavior to match.

**Environment:**
 - CockroachDB version v19.1.5 and v19.2
 - Client app: customers go driver and cockroach sql

**Comment by RoachietheSupportRoach** [comment]
Zendesk ticket #[4111](https://cockroachdb.zendesk.com/agent/tickets/4111) has been linked to this issue.

**Comment by jordanlewis** [comment]
I think this might be basically the same root cause as the halloween problem.

@justinj what do you think? What snapshot of the data is used by other parts of the query, after a CTE executes a DELETE? Is that a different snapshot than what Postgres uses?

**Comment by justinj** [comment]
yes, statements should not be able to see their own mutations but today Cockroach does not guarantee this. my expectation/hope is that this will be fixed by the same change that resolves the halloween problem.

**Comment by knz** [comment]
fixed in 20.1
